plugins/videobasedtracker/UVCImageSource.cpp

```diff
@@ -173,12 +176,16 @@ namespace vbtracker {
 
     bool UVCImageSource::grab()
     {
-        // libuvc doesn't provide for separate grab() and retrieve() operations.
-        // Instead, grab() will set a flag to alert the callback function that
-        // we wish to store the next frame and retrieve() will return that
-        // frame.
-        std::lock_guard<std::mutex> guard(mutex_);
-        return (frame_ != nullptr);
+      //Gain access to frames_ queue
+      std::unique_lock<std::mutex> lock(mutex_);
+
+      //Check if there are any frames available
+      if (frames_.empty())
+	//Wait for some frames to become available
+	frames_available_.wait(lock);
```

Comment:
I only use one notify statement, which happens after the frame is added, but I agree, a predicated version would be neater. I've pushed a patch for this.